crypto/src/lib.rs

```diff
@@ -274,89 +299,189 @@ impl From<KeyPair> for (PublicKey, PrivateKey) {
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl fmt::Debug for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple(self.algorithm().as_static_str())
+            .field(&self.normalize())
+            .finish()
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl fmt::Display for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.normalize())
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl core::hash::Hash for PublicKey {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        (self.algorithm(), self.payload()).hash(state)
+    }
+}
+
+impl PartialOrd for PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PublicKey {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        (self.algorithm(), self.payload()).cmp(&(other.algorithm(), other.payload()))
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl Encode for PublicKey {
+    fn size_hint(&self) -> usize {
+        self.algorithm().size_hint() + self.payload().size_hint()
+    }
+
+    fn encode_to<W: parity_scale_codec::Output + ?Sized>(&self, dest: &mut W) {
+        self.algorithm().encode_to(dest);
+        self.payload().encode_to(dest);
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
+impl Decode for PublicKey {
+    fn decode<I: parity_scale_codec::Input>(
+        input: &mut I,
+    ) -> Result<Self, parity_scale_codec::Error> {
+        let algorithm = Algorithm::decode(input)?;
+        let payload = Vec::decode(input)?;
+        Self::from_raw(algorithm, &payload).map_err(|_| {
+            parity_scale_codec::Error::from(
+                "Failed to construct public key from digest function and payload",
+            )
         })
     }
 }
 
+#[cfg(not(feature = "ffi_import"))]
+impl IntoSchema for PublicKey {
+    fn type_name() -> String {
+        Self::id()
+    }
+
+    fn update_schema_map(metamap: &mut MetaMap) {
+        if !metamap.contains_key::<Self>() {
+            if !metamap.contains_key::<Algorithm>() {
+                <Algorithm as iroha_schema::IntoSchema>::update_schema_map(metamap);
+            }
+            if !metamap.contains_key::<ConstVec<u8>>() {
+                <ConstVec<u8> as iroha_schema::IntoSchema>::update_schema_map(metamap);
+            }
+
+            metamap.insert::<Self>(Metadata::Struct(NamedFieldsMeta {
+                declarations: vec![
+                    Declaration {
+                        name: String::from("algorithm"),
+                        ty: core::any::TypeId::of::<Algorithm>(),
+                    },
+                    Declaration {
+                        name: String::from("payload"),
+                        ty: core::any::TypeId::of::<ConstVec<u8>>(),
+                    },
+                ],
+            }));
+        }
+    }
+}
+
+#[cfg(not(feature = "ffi_import"))]
```

Comment:
```suggestion
```

this is not necessary?

Reply:
also, you can derive `TypeId`

Reply:
> this is not necessary?

It still kind of not obvious to me which parts should be under this feature-gate

Reply:
the ones that operate on fields of opaque structs